### PR TITLE
feat: impl'd describe for pods (#553)

### LINKED ISSUE #553: provide "describe" like in vscode
taken from https://github.com/redhat-developer/intellij-kubernetes/pull/548#issuecomment-1399406965
CC @cloorc

"Describe" is provided in the context menu of resource tree items in vscode. 
image
It prints detailed informations about a k8s resources.
image

### PR DESCRIPTION
**Pull request by adietish**
fixes #553 

How to test this:

This implementation should provide the same output as `kubectl describe pod <name>`. 

**Steps:**
1. EXEC: copy & paste the following into an editor in IJ, naming the file "describe-test.yml": 
https://gist.github.com/adietish/b9bde277776f657af7d76515effa1882#file-describe-test-yml. Then push the editor to the cluster.
(alternatively you can also use kubectl with the file: `kubectl apply -f describe-test.yml`)
2. ASSERT: a `Pod` **describe-test** is created and show up in the resource tree 
3. EXEC: pick "Describe" in the context menu to the pod **describe-test** ([cluster] > Workloads > Pods > describe-test)
4. ASSERT: An editor is opened describing the pod. The title of the editor is "Describe Pod describe-test".
5. EXEC: run `kubectl describe pod describe-test` and compare the output with the content of the editor

**Comparing the outputs:**
You can use a diff or https://www.diffchecker.com/text-compare/ which helps a lot.
There are limitations that were not implemented yet (ex. the list of events that kubectl is printing in the end of the output):
* #778
* #774
* #776
* #777

The following examples show equivalence in intellij-kubernetes and kubectl. The differences are mainly in formatting, cause by the fact the intellij-kubernetes is producing YAML whereas `kubectl` is simply printing text.

***Examples***: 
* `kubectl`
```
Labels:     luke=skywalker
            princess=leia
```
*  `intellij-kubernetes` (creates a yaml mapping, `<key>:<value>` instead of `<key>=<value>`)
```yaml
Labels:
  luke: skywalker
  princess: leia
```

* `kubectl`
```
Medium:
```
* `intellij-kubernetes` (prints `<none>` instead of ``)
```yaml
Medium: <none>
```
* `kubectl`
```
Readiness Gates:
  Type                     Status
  www.example.com/gate-1   <none>
  ContainersReady          <none>
```
* `intellij-kubernetes` (skips heading: "Type       Status")
```
  www.example.com/gate-1: <none>
  ContainersReady: <none>
```

* `kubectl`
```
Tolerations:    gpu=true:NoSchedule
                node.kubernetes.io/not-ready:NoExecute for 3600s
                node.kubernetes.io/unreachable:NoExecute op=Exists for 3600s
                project=system:NoSchedule
```
* `intellij-kuberetes` (creates a yaml sequence, each element starts with `-`):
```yaml
Tolerations:
- gpu=true:NoSchedule
- project=system:NoSchedule
- node.kubernetes.io/unreachable:NoExecute op=Exists for 3600s
- node.kubernetes.io/not-ready:NoExecute for 3600s
```

* `kubectl`
```
Environment:
      MY_POD_NAME:             describe-test (v1:metadata.name)
```
* `intellij-kubernetes` (does not resolve expression `v1:metadata.name` to the value `describe-test`. It only prints the expression `(v1:metadata-name)`. Known limitation, filed to #774)
```yaml
Environment:
    MY_POD_NAME: (v1:metadata.name)
```

### CODE REVIEW

**Comment by adietish** [comment]
@sbouchet, @olkornii: please test/review this. Instructions are in the 1st comment to this issue. When compared to `kubectl describe` there are limitations. Beside those, this impl should produce equivalent output in yaml.
Thanks!

**Comment by sbouchet** [comment]
issue discovered when describing a simple pod : 
```
---
apiVersion: "v1"
kind: "Pod"
metadata:
  labels:
    jedi: "luke"
  name: "alpine1"
spec:
  containers:
    - image: "alpine"
      name: "alpine"
      ports:
        - containerPort: 8080
```

the kubectl describe returns  
`Node:             minikube/192.168.39.85`

while intellij-kubes returns 
`Node: minikube/10.244.3.78`

**Comment by sbouchet** [comment]
issues discovered while describing a pod created with helm chart ( kuberos ):
`kubectl`
```
Name:             kube-kuberos-659f864687-hd889
Namespace:        myproj
Priority:         0
Service Account:  default
Node:             minikube/192.168.39.85
Start Time:       Mon, 15 Jul 2024 17:48:18 +0200
Labels:           app=kuberos
                  checksum/config=2c10ca940734118cab9c36c72a4b8e7c49570b8d5022d8636e780df38d43a74
                  pod-template-hash=659f864687
                  release=kube
Annotations:      checksum/config: 2c10ca940734118cab9c36c72a4b8e7c49570b8d5022d8636e780df38d43a74
Status:           Running
IP:               10.244.3.87
IPs:
  IP:           10.244.3.87
Controlled By:  ReplicaSet/kube-kuberos-659f864687
Containers:
  kuberos:
    Container ID:  docker://2be868e2f251764998e133426963e2ae866c806d85d183f40925b2297a3a51b5
    Image:         negz/kuberos:ede4085
    Image ID:      docker-pullable://negz/kuberos@sha256:5fce210b3a813083da0ef60ac10b7cdf81f274c866b92598f3d07d6aa4ff4d2e
    Port:          10003/TCP
    Host Port:     0/TCP
    Command:
      /kuberos
      https://accounts.google.com
      REDACTED.apps.googleusercontent.com
      /cfg/secret
      /cfg/template
    State:          Running
      Started:      Tue, 30 Jul 2024 10:08:55 +0200
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Tue, 30 Jul 2024 10:08:25 +0200
      Finished:     Tue, 30 Jul 2024 10:08:54 +0200
    Ready:          True
    Restart Count:  11
    Liveness:       http-get http://:kuberos-port/healthz delay=0s timeout=1s period=10s #success=1 #failure=3
    Readiness:      http-get http://:kuberos-port/healthz delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:    <none>
    Mounts:
      /cfg/secret from secret (rw,path="oidcSecret")
      /cfg/template from template (rw,path="template")
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-mlnjq (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       True 
  ContainersReady             True 
  PodScheduled                True 
Volumes:
  template:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      kube-kuberos-config
    Optional:  false
  secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kuberos-secret
    Optional:    false
  kube-api-access-mlnjq:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>
```
`intellij-kubernetes`
```
Name: kube-kuberos-659f864687-hd889
Namespace: myproj
Priority: 0
Service Account: default
Node: minikube/10.244.3.87
Start Time: Mon, 15 Jul 2024 15:48:18 +0200
Labels:
  app: kuberos
  checksum/config: 2c10ca940734118cab9c36c72a4b8e7c49570b8d5022d8636e780df38d43a74
  pod-template-hash: 659f864687
  release: kube
Annotations:
  checksum/config: 2c10ca940734118cab9c36c72a4b8e7c49570b8d5022d8636e780df38d43a74
Status: Running
IP: 10.244.3.87
IPs:
  IP: 10.244.3.87
Controlled By: ReplicaSet/kube-kuberos-659f864687
Init Containers: <none>
Containers:
  kuberos:
    Container ID: docker://2be868e2f251764998e133426963e2ae866c806d85d183f40925b2297a3a51b5
    Image: negz/kuberos:ede4085
    Image ID: docker-pullable://negz/kuberos@sha256:5fce210b3a813083da0ef60ac10b7cdf81f274c866b92598f3d07d6aa4ff4d2e
    Port: 10003/TCP
    Host Port: 0/TCP
    Command:
    - /kuberos
    - https://accounts.google.com
    - REDACTED.apps.googleusercontent.com
    - /cfg/secret
    - /cfg/template
    State: Running
    Started: Tue, 30 Jul 2024 08:08:55 +0200
    Last State: Terminated
    Reason: Error
    Exit Code: 2
    Finished: Tue, 30 Jul 2024 08:08:54 +0200
    Ready: true
    Restart Count: 11
    Liveness: 'http-get http:/healthz delay=nulls timeout=1s period=10s #success=1
      #failure=3'
    Readiness: 'http-get http:/healthz delay=nulls timeout=1s period=10s #success=1
      #failure=3'
    Environment: <none>
    Mounts:
    - /cfg/template from template (rw, path = "template")
    - /cfg/secret from secret (rw, path = "oidcSecret")
    - /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-mlnjq (ro)
Conditions:
  PodReadyToStartContainers: true
  Initialized: true
  Ready: true
  ContainersReady: true
  PodScheduled: true
Volumes:
  template:
    Type: ConfigMap (a volume populated by a ConfigMap)
    Name: kube-kuberos-config
    Optional: <none>
  secret:
    Type: Secret (a volume populated by a Secret)
    SecretName: kuberos-secret
    Optional: <none>
  kube-api-access-mlnjq:
    Type: Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds: '3607'
    ConfigMapName: kube-root-ca.crt
    ConfigMapOptional: <none>
    DownwardAPI: 'true'
QoS Class: BestEffort
Node-Selectors: <none>
Tolerations:
- node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
- node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
```

note the `Node: ` issue same as above
then on the Containers part, there is missing spaces for `State / Started  ` and `Last State / Reason...`
on the same place, there is a missing `Started` value
later on the `liveness/readiness` seems delay is returning null ( should be 0 )

On `Volumes `, the `Optional` value should be `False` ( returns <none> )

**Comment by adietish** [comment]
@sbouchet thx for the review. On top of your issues I saw the following:
* kubectl:
```
TokenExpirationSeconds: 3607
```
* intellij-kubernetes
```
TokenExpirationSeconds: '3607'
```


* kubectl:
```
DownwardAPI: true
```
* intellij-kubernetes
```
DownwardAPI: 'true'
```

**Comment by adietish** [comment]
@sbouchet please re-review, I have fixed the issues that you have found. Thanks!

**Comment by adietish** [comment]
@sbouchet I also changed `Command` to better match kubectl:
* kubectl:
```bash
   Command:
    - /kuberos
    - https://accounts.google.com
    - REDACTED.apps.googleusercontent.com
    - /cfg/secret
    - /cfg/template
```
* intellij-kubernetes:
```yaml
   Command: |-
    - /kuberos
    - https://accounts.google.com
    - REDACTED.apps.googleusercontent.com
    - /cfg/secret
    - /cfg/template
```

**Comment by adietish** [comment]
@sbouchet please re-review, thanks!

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/actions/DescribeResourceAction.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Red Hat, Inc.
```
Comment (sbouchet) [inline src/main/kotlin/com/redhat/devtools/intellij/kubernetes/actions/DescribeResourceAction.kt:2]:
```suggestion
 * Copyright (c) 2024 Red Hat, Inc.
```

Comment (sbouchet) [reply src/main/kotlin/com/redhat/devtools/intellij/kubernetes/actions/DescribeResourceAction.kt:2]: nitpicking

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/editor/KubernetesEditorsTabTitleProvider.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
```
Comment (sbouchet) [inline src/main/kotlin/com/redhat/devtools/intellij/kubernetes/editor/KubernetesEditorsTabTitleProvider.kt:2]:
```suggestion
 * Copyright (c) 2024 Red Hat, Inc.
```

File: src/test/kotlin/com/redhat/devtools/intellij/kubernetes/model/mocks/PodContainer.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package com.redhat.devtools.intellij.kubernetes.model.mocks
```
Comment (sbouchet) [inline src/test/kotlin/com/redhat/devtools/intellij/kubernetes/model/mocks/PodContainer.kt:1]: please add copyright header

File: src/test/kotlin/com/redhat/devtools/intellij/kubernetes/model/util/PodUtilsTest.kt [hunk 4]
```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Red Hat, Inc.
```
Comment (sbouchet) [inline src/test/kotlin/com/redhat/devtools/intellij/kubernetes/model/util/PodUtilsTest.kt:2]:
```suggestion
 * Copyright (c) 2024 Red Hat, Inc.
```

**Review by sbouchet — APPROVED** [review]
nitpicking copyrights. LGTM !

**Pull request merged**